app.js

```diff
@@ -483,16 +587,95 @@ const completeRecoveryValidation = ({ isNew } = {}) => async ctx => {
     ctx.body = await recoveryMethodsFor(account);
 };
 
+const completeRecoveryValidation = ({ isNew } = {}) => async (ctx) => {
+    const {
+        accountId,
+        method,
+        securityCode,
+        enterpriseRecaptchaToken,
+        recaptchaAction,
+        recaptchaSiteKey
+    } = ctx.request.body;
+
+    if (!securityCode || isNaN(parseInt(securityCode, 10)) || securityCode.length !== 6) {
+        ctx.throw(401, 'valid securityCode required');
+    }
+
+    const account = await AccountService.getAccount(accountId);
+    if (!account) {
+        ctx.throw(401, 'account does not exist');
+    }
+
+    const [recoveryMethod] = await RecoveryMethodService.listRecoveryMethods({
+        accountId,
+        detail: method.detail,
+        kind: method.kind,
+        securityCode,
+    });
+
+    if (!recoveryMethod) {
+        ctx.throw(401, 'recoveryMethod does not exist');
+    }
+
+    // for new accounts, clear all other recovery methods that may have been created
+    if (isNew) {
+        await RecoveryMethodService.deleteOtherRecoveryMethods({ accountId, detail: method.detail });
+    }
+
+    await RecoveryMethodService.updateRecoveryMethod({
```

Comment:
Looks like the update() call in the legacy function only set the one value (securityCode)... is providing the detail/kind here redundant or could this in some way change our existing behaviour?

Reply:
This is one of the things that makes this conversion so hard to reason about 🙁 

The fetch for `recoveryMethod` in the legacy method is:
```
const [recoveryMethod] = await account.getRecoveryMethods({
    where: {
        kind: method.kind,
        detail: method.detail,
        securityCode: securityCode
    }
});

...

await recoveryMethod.update({ securityCode: null });
```

Since `recoveryMethod` necessarily has `detail === method.detail` and `kind === method.kind` already, updating _should_ be a no-op. Then in `/sequelize/recovery_method.js`, the `updateRecoveryMethod` method does the same fetch anyway:
```
async updateRecoveryMethod({ accountId, detail, kind, securityCode }) {
    const [recoveryMethod] = await this.listRecoveryMethods_internal({ accountId, detail, kind, securityCode });
    ...
}
```

which ensures that it's still fetching the same thing even if it is doing a redundant update. It's possible that the fetch is nondeterministic if another `detail` + `kind` pair existed with a different `publicKey` (the combination of which constitutes a unique tuple based on the index), but since it just takes the first value in the destructure anyway, the logic should be identical.

It's still a bug however, since I'm not actually setting `securityCode = null` on line 625 🤦. So you might want to double check my explanation once more to make sure it parses 😄